cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,162 @@
+      String metadataTableName = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAME);
+      String hbaseNamespacePrefix = tableDesc.getValue(Constants.Dataset.TABLE_PREFIX);
+      HTableNameConverter nameConverter = new HTable98NameConverter();
+      metadataTable = env.getTable(nameConverter.toTableName(hbaseNamespacePrefix,
+                                                             TableId.from(metadataTableNamespace, metadataTableName)));
+      topicCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(2, TimeUnit.MINUTES)
+        .maximumSize(1000)
+        .build(new CacheLoader<TopicId, TopicMetadata>() {
+
+          @Override
+          public TopicMetadata load(TopicId topicId) throws Exception {
+            Get get = new Get(MessagingUtils.toRowKeyPrefix(topicId));
+            Result result = metadataTable.get(get);
+            byte[] properties = result.getValue(COL_FAMILY, COL);
+            Map<String, String> propertyMap = GSON.fromJson(Bytes.toString(properties), MAP_TYPE);
+            return new TopicMetadata(topicId, propertyMap);
+          }
+        });
+    }
+  }
+
+  @Override
+  public InternalScanner preFlushScannerOpen(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
+                                             KeyValueScanner memstoreScanner, InternalScanner s) throws IOException {
+    LOG.info("preFlush, filter using TTLFilter");
+    Scan scan = new Scan();
+    scan.setFilter(new TTLFilter(c.getEnvironment(), System.currentTimeMillis()));
+    return new StoreScanner(store, store.getScanInfo(), scan, Collections.singletonList(memstoreScanner),
+                            ScanType.COMPACT_DROP_DELETES, store.getSmallestReadPoint(), HConstants.OLDEST_TIMESTAMP);
+  }
+
+  @Override
+  public InternalScanner preCompactScannerOpen(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
+                                               List<? extends KeyValueScanner> scanners, ScanType scanType,
+                                               long earliestPutTs, InternalScanner s,
+                                               CompactionRequest request) throws IOException {
+    LOG.info("preCompact, filter using TTLFilter");
+    Scan scan = new Scan();
+    scan.setFilter(new TTLFilter(c.getEnvironment(), System.currentTimeMillis()));
+    return new StoreScanner(store, store.getScanInfo(), scan, scanners, scanType, store.getSmallestReadPoint(),
+                            earliestPutTs);
+  }
+
+  private final class TTLFilter extends FilterBase {
+    private final RegionCoprocessorEnvironment env;
+    private final long timestamp;
+
+    TTLFilter(RegionCoprocessorEnvironment env, long timestamp) {
+      this.env = env;
+      this.timestamp = timestamp;
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell cell) throws IOException {
+      byte[] rowKey = Arrays.copyOfRange(cell.getRow(), prefixLength, cell.getRowLength());
```

Comment:
Better not to create new array as this is happening for each individual cell during flush/compact time. This means we might be not using the `ImmutableMessageTableEntry`, but rather just parse out the information we need. We might need to refactor the code a bit to have the parsing logic becomes helper method on `MessagingUtils` and coprocessor class only uses `MessagingUtils` and move that class to `hbase-compat-base` (it's not an ideal place, but at least it's a common parent between compat modules and cdap-tms)